### PR TITLE
[DOCS] Testing ADR

### PR DESCRIPTION
**Pull request by donaldheppner**
ADR describing our approach to automated testing.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code

### CODE REVIEW

File: docs/adr/0003-automated-testing.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+>
+> Tests that do these things aren't bad. Often they are worth writing, and you generally will write them in unit test harnesses. However, it is important to be able to separate them from true unit tests so that you can keep a set of tests that you can run _fast_ whenever you make changes.
+
+## Decision
+
+All unit tests must be marked with `@pytest.mark.unit`, other tests must be marked with `@pytest.mark.integration`. To be considered unit tests, they must meet the above defintion.
+
+## Consequences
+
+Doing this effectively requires both that we instrument our build pipeline to detect improperly marked tests and an ongoing effort to tag our existing test suites. As we work to increase the isolation necessary to localize problems and keep tests fast, this effort will drive us towards new coding patterns (such as dependency injection).
```
Comment (Kilo59) [inline docs/adr/0003-automated-testing.md:38]: ❤️

**Review by Kilo59 — APPROVED** [review]

File: docs/adr/0003-automated-testing.md (hunk 1 above)
Comment (alexsherstinsky) [reply docs/adr/0003-automated-testing.md:38]: @donaldheppner Not sure why tests are failing.  Also, what does ADR stand for (sorry if I am missing the obvious).  Thank you.

Comment (alexsherstinsky) [reply docs/adr/0003-automated-testing.md:38]: Nevermind: ADR - "Architecture Decision Record" (for others like me who are not so good with the acronyms).

**Review by anthonyburdi — APPROVED** [review]
Love this. Thank you for writing this up! I don't see a direct way to enforce tags via pytest itself - if there is a library (ruff maybe?) that we can use off the shelf to polish and ratchet (cc @Kilo59) that would be great, otherwise we can always parse the source code (ast).

**Comment by Kilo59** [comment]
@anthonyburdi 

There is a set of ruff rules for `flake8-pytest-style`, it has a lot of good rules that could help us but I don't see anything about enforcing the use of a particular marker. 😢 
https://beta.ruff.rs/docs/rules/#flake8-pytest-style-pt

**Pull request merged**